### ISSUE TITLE
DEPR: pandas/core

### ISSUE THREAD

**Issue by jreback** [pandas-dev/pandas#27522]
we have a documentation note that pandas.core is private but we need an actual deprecation

**Comment by WillAyd** [comment]
Is there anything outside of the top level pandas namespace that is public? I know we have some usages of `pandas.io.json.json_normalize` and IIRC something with Categorical but I wonder if we should just move them to `pandas` and provide a generic warning for anything not coming from `pandas` to end users

**Comment by jbrockmendel** [comment]
I share Will's sentiment, though from looking at the history it seems like there was an intentional move away from that before I got here.

If there is a shuffling of core files, I'd like to separate out Index/Series/DataFrame/internals from lower-level code that doesn't need to be "aware" of these classes

**Comment by jreback** [comment]
nothing should be public except pd namespace
but historically we had a lot (like everything) was

we moved everything a few years ago to create _libs and consolidate under core & io
but i think we need to be explicitly private

moving things around is much much easier after

**Comment by WillAyd** [comment]
So sounds like we need to:

  1. Audit external usage of anything not in the top level (json_normalize, etc...) and make available in pandas.__init__.py
  2. Document / throw FutureWarning when accessing other namespaces

Then actually prefix those namespaces with an underscore to privatize?

**Comment by huandzh** [comment]
Hi, it will be nice to also expose exceptions raised by top level functions, since downstream may want to handle them in their own code.

Use case: https://github.com/aliyun/aliyun-odps-python-sdk/pull/99

**Comment by simonjayhawkins** [comment]
> Hi, it will be nice to also expose exceptions raised by top level functions, since downstream may want to handle them in their own code.

this may warrant a separate issue for discussion. @jorisvandenbossche what's you current feeling regarding custom exceptions exposed to users? https://github.com/pandas-dev/pandas/pull/25569#issuecomment-471495727

**Comment by jreback** [comment]
> > Hi, it will be nice to also expose exceptions raised by top level functions, since downstream may want to handle them in their own code.
> 
> this may warrant a separate issue for discussion. @jorisvandenbossche what's you current feeling regarding custom exceptions exposed to users? [#25569 (comment)](https://github.com/pandas-dev/pandas/pull/25569#issuecomment-471495727)

these are already mostly in pandas.exceptions which are public

**Comment by simonjayhawkins** [comment]
> > > Hi, it will be nice to also expose exceptions raised by top level functions, since downstream may want to handle them in their own code.
> > 
> > 
> > this may warrant a separate issue for discussion. @jorisvandenbossche what's you current feeling regarding custom exceptions exposed to users? [#25569 (comment)](https://github.com/pandas-dev/pandas/pull/25569#issuecomment-471495727)
> 
> these are already mostly in pandas.exceptions which are public

is this an exception (sorry for the pun) to the "nothing should be public except pd namespace" rule?

**Comment by jreback** [comment]
> > > > Hi, it will be nice to also expose exceptions raised by top level functions, since downstream may want to handle them in their own code.
> > > 
> > > 
> > > this may warrant a separate issue for discussion. @jorisvandenbossche what's you current feeling regarding custom exceptions exposed to users? [#25569 (comment)](https://github.com/pandas-dev/pandas/pull/25569#issuecomment-471495727)
> > 
> > 
> > these are already mostly in pandas.exceptions which are public
> 
> is this an exception (sorry for the pun) to the "nothing should be public except pd namespace" rule?

yes

**Comment by jreback** [comment]
pandas.api as well; the other top level packages should be private (though may be tricky / take a while to fully remove)

**Comment by TomAugspurger** [comment]
Since we're re-having this discussion, I'll note that if we're disciplined
about making private modules private with a leading underscore,
there's no need for a `.api` namespace. Everything not starting with an
underscore is public.

pandas.arrays
pandas.exceptions
pandas.extensions
pandas.types
...

Tom

**Comment by huandzh** [comment]
> 
> 
> > > > > Hi, it will be nice to also expose exceptions raised by top level functions, since downstream may want to handle them in their own code.
> > > > 
> > > > 
> > > > this may warrant a separate issue for discussion. @jorisvandenbossche what's you current feeling regarding custom exceptions exposed to users? [#25569 (comment)](https://github.com/pandas-dev/pandas/pull/25569#issuecomment-471495727)
> > > 
> > > 
> > > these are already mostly in pandas.exceptions which are public
> > 
> > 
> > is this an exception (sorry for the pun) to the "nothing should be public except pd namespace" rule?
> 
> yes

Then, `OptionError` raised by `pandas.set_option` will be available in `pandas.exceptions` afterwards, and so do other custom exceptions raised by top-level functions, am i right?

Anyway, it's better to have them in one and fixed place.

**Comment by simonjayhawkins** [comment]
> > > > > > Hi, it will be nice to also expose exceptions raised by top level functions, since downstream may want to handle them in their own code.
> > > > > 
> > > > > 
> > > > > this may warrant a separate issue for discussion. @jorisvandenbossche what's you current feeling regarding custom exceptions exposed to users? [#25569 (comment)](https://github.com/pandas-dev/pandas/pull/25569#issuecomment-471495727)
> > > > 
> > > > 
> > > > these are already mostly in pandas.exceptions which are public
> > > 
> > > 
> > > is this an exception (sorry for the pun) to the "nothing should be public except pd namespace" rule?
> > 
> > 
> > yes
> 
> Then, `OptionError` raised by `pandas.set_option` will be available in `pandas.exceptions` afterwards, and so do other custom exceptions raised by top-level functions, am i right?
> 
> Anyway, it's better to have them in one and fixed place.

it looks like this was added in #19790 and intended to be user facing. so it probably should be in the public api. can you raise an issue for this.

**Comment by huandzh** [comment]
> 
> 
> > > > > > > Hi, it will be nice to also expose exceptions raised by top level functions, since downstream may want to handle them in their own code.
> > > > > > 
> > > > > > 
> > > > > > this may warrant a separate issue for discussion. @jorisvandenbossche what's you current feeling regarding custom exceptions exposed to users? [#25569 (comment)](https://github.com/pandas-dev/pandas/pull/25569#issuecomment-471495727)
> > > > > 
> > > > > 
> > > > > these are already mostly in pandas.exceptions which are public
> > > > 
> > > > 
> > > > is this an exception (sorry for the pun) to the "nothing should be public except pd namespace" rule?
> > > 
> > > 
> > > yes
> > 
> > 
> > Then, `OptionError` raised by `pandas.set_option` will be available in `pandas.exceptions` afterwards, and so do other custom exceptions raised by top-level functions, am i right?
> > Anyway, it's better to have them in one and fixed place.
> 
> it looks like this was added in #19790 and intended to be user facing. so it probably should be in the public api. can you raise an issue for this.

#27553 submitted

**Comment by jorisvandenbossche** [comment]
> nothing should be public except pd namespace

I think it has become clear from the discussion, but to put it explicitly: this is not really correct (in the meaning of "only objects and functions in the pd namespace").

All submodules in the pd namespace are also (partly) public (and the objects they contain), *except* for `pandas.core`, `pandas.util` and `pandas.compat`. For `pandas.io` and `pandas.tseries` this is only "partly", as we say in the docs that for those submodules only those objects are public which are listed in the API docs. But so the `arrays`, `errors`, `plotting` and `testing` modules are fully public. 
See the docs on this: https://pandas.pydata.org/pandas-docs/stable/reference/index.html 

I think the main confusing situation are those submodules that are only partly public (io and tseries), the objects of them we list in the API docs as public:

* pandas.io:
  * mentioned in the API reference:
    * pandas.io.json: `json_normalize`, `build_table_schema`
    * pandas.io.formats.style: `Styler`
  * also used in the user guide: `pd.io.sql.execute` (and pd.io.sql.get_schema always had a somewhat dubious status) and `pd.io.date_converters.parse_date_time`
* pandas.tseries:
  * mentioned in the API reference:
    * pandas.tseries.offsets: `DateOffset`, `Tick`, and all their subclasses
    * pandas.tseries.frequencies: `to_offset`
  * also used in the user guide: several objects from `pandas.tseries.holiday` 
  * Additional note: in the user guide we use `pd.offsets` and not `pd.tseries.offsets` as in the reference guide, something we should make consistent.

To say that `json_normalize` is far from the only non-top-level function.

From the explicitly documented as private modules (core, compat, util), I think it is mainly `pd.util.testing` that actually sees quite some usage in the wild. Not sure if we want to do something about this. We moved some to `pandas.testing`, but I think people use more than what we exposed there (eg several "create dataframe" functions).

For the custom expections that are not included in `pandas.errors`, I opened https://github.com/pandas-dev/pandas/issues/27656 (there are quite a few).

**Comment by TomAugspurger** [comment]
I don't have a strong preference for what we do, but I would like to be
consistent. So we shouldn't move just `pd.io.json.json_normalize` to the
top-level, but not `pd.io.json.build_table_schema`.

**Comment by TomAugspurger** [comment]
Is deprecating this a priority for 1.0? According to https://dev.pandas.io/docs/development/policies.html, we're allowed to add depredations in x.y releases, so we can do this after 1.0 if no one has time for it.

**Comment by jorisvandenbossche** [comment]
It's not fully clear to me what is actually being proposed here (a lot of the above discussion was about other top-level modules that are actually public). Is it about moving `pandas.core` to eg `pandas._core` and raise warnings for `pandas.core` ?

**Comment by jreback** [comment]
> It's not fully clear to me what is actually being proposed here (a lot of the above discussion was about other top-level modules that are actually public). Is it about moving `pandas.core` to eg `pandas._core` and raise warnings for `pandas.core` ?

yes that’s what we did quite a while ago for the other codebases

**Comment by TomAugspurger** [comment]
Just FYI, we do have public things currently exposed from `pandas.core`, things like Resampler, GroupBy, etc. If we're deprecating all of `pandas.core` then we'll need to add those elsewhere and update the API references.

**Comment by jorisvandenbossche** [comment]
Moving this off the 1.0 milestone (deprecations are not blockers for 1.0 I think, since they can be introduced in any 1.x release)

**Comment by lithomas1** [comment]
take

**Comment by jorisvandenbossche** [comment]
@lithomas1 I think this warrants more discussion first.

Personally, I am not sure it is worth it to go through a renaming / deprecation. I think a "everything in pandas.core is private" is also a fine message.

**Comment by TomAugspurger** [comment]
This could probably use a decision. I think it'll require a DeprecationWarning -> FutureWarning path like we do for other large changes. And we'll want to leave projects plenty of time to update to this large of a change.

Regardless, we should be changing where we publicly expose things from `pandas.core`, e.g. `DataFrameGroupby`, `Rolling`, etc. Those should be exposed in a public namespace as a precursor.

**Comment by jreback** [comment]
i think we should do it and show a DeprecationWarning

**Comment by simonjayhawkins** [comment]
removing milestone

**Comment by rhshadrach** [comment]
I was thinking this was blocked since `Styler` is not in `pandas.api.typing`, but `Styler` is defined in `pandas.io` and not `pandas.core`.

@phofl - My recollection is that you had better ideas on how to implement this than I could conjure. Any interest in tackling this at some point as part of 2.x?

**Comment by rhshadrach** [comment]
> * It's certainly not the case for all of those examples, but for some of them there might be a good reason that a downstream project is using those APIs. And then the question is: do we want to expose some of those in a public location? (and where?) Or do we just assume that those projects will keep using them from `pandas._core` (but then we risk many people simply adding try/excepts first importing from `pandas._core` falling back to importing from `pandas.core`, and then what did we achieve with this?)

Since we're doing the deprecation very slowly (over 2+ years maybe?), there will be an extended period of time where you can import from either `_core` or `core`. By the time we remove `core`, I expect packages will only be supporting versions of pandas that have `_core`. If this is the case, there is no need for try-excepts.

With this, I think it's okay to think of this question as independent of our deprecation of `core`.

**Comment by lithomas1** [comment]
IMO, this probably should require a PDEP, given the scope of this deprecation.

I missed the last dev call, was something like that discussed there?

**Comment by rhshadrach** [comment]
> IMO, this probably should require a PDEP, given the scope of this deprecation.

pandas.core is documented as being private. We're deprecating the move because users haven't always followed this, but in my opinion, this isn't technically a user-facing change.

**Comment by lithomas1** [comment]
Even though this doesn't directly affect users, I think it would be good to provide clarity to downstream as to how they should go about replacing the internals from pandas.core they are using, and the timeline for doing so. Making this a PDEP would provide a lot more visibility than just discussing on the issue here.

There is also a large impact on contributors to the package as I imagine stuff like renaming ``core`` to ``_core`` will cause a bunch of conflicts, and maybe mess up the git history (not sure if git blame is able to handle this or not).

**Comment by rhshadrach** [comment]
@pandas-dev/pandas-core - who thinks this does or doesn't need a PDEP.

:+1: Needs a PDEP
:-1: Doesn't need a PDEP.

**Comment by Dr-Irv** [comment]
> > IMO, this probably should require a PDEP, given the scope of this deprecation.
> 
> pandas.core is documented as being private. We're deprecating the move because users and downstream packages haven't always followed this, but in my opinion, this isn't technically a user-facing change.

Yes, but we do have reference pages that list methods as part of `core`.  If you go to https://pandas.pydata.org/pandas-docs/stable/reference/groupby.html and then click on any of the `DataFrameGroupBy` methods, they are showing as being part of `core` .

**Comment by Dr-Irv** [comment]
> Even though this doesn't directly affect users, I think it would be good to provide clarity to downstream as to how they should go about replacing the internals from pandas.core they are using, and the timeline for doing so. Making this a PDEP would provide a lot more visibility than just discussing on the issue here.

To me, this is a compelling reason to do so.  Then we have an agreed upon path for the deprecation.  I don't think we need a PDEP that is about renaming.  We need a PDEP that explains the deprecation process.

**Comment by phofl** [comment]
> We need a PDEP that explains the deprecation process.

That's what docs are there for

**Comment by Dr-Irv** [comment]
> > We need a PDEP that explains the deprecation process.
> 
> That's what docs are there for

Yes, but I think we have to agree on that deprecation process before it gets documented (effectively making the core of the PDEP the future documentation), as opposed to doing a PR that has a description of the deprecation process in docs, which we then have a debate over.

**Comment by mroeschke** [comment]
I don't think this needs a PDEP (yet) because:

1. There already seems to be general consensus from the core team to deprecate `pandas.core`
2. There hasn't been objection to the plan proposed in https://github.com/pandas-dev/pandas/issues/27522#issue-471168284 (due to lack of discussion it appears)

From the comments so far it appears the deprecation plan needs more clarity/details(?). If the plan gets to a state where there's a gridlock in consensus then I think a PDEP vote would be warranted. Also IMO the best way to get feedback on this procedure from users would be to issue to deprecation as opposed to writing a PDEP

**Comment by rhshadrach** [comment]
@Dr-Irv 

> but I think we have to agree on that deprecation process before it gets documented

That's what this issue is for.

**Comment by Dr-Irv** [comment]
> That's what this issue is for. I've edited the OP here with the process as I currently see it, but will be updating based on feedback.

Fair enough.  Your edits cleared things up.  I think the section at the end entitled "Anytime after" could use more precision on the timing.  Would those changes in that section be made in a major or minor release?
